src/core_plugins/elasticsearch/lib/create_proxy.js

```diff
@@ -17,60 +17,63 @@
  * under the License.
  */
 
-import createAgent from './create_agent';
-import mapUri from './map_uri';
-import { assign } from 'lodash';
+import Joi from 'joi';
+import Boom from 'boom';
 
-export function createPath(prefix, path) {
-  path = path[0] === '/' ? path : `/${path}`;
-  prefix = prefix[0] === '/' ? prefix : `/${prefix}`;
+export function createProxy(server) {
+  const { callWithRequest } =  server.plugins.elasticsearch.getCluster('data');
 
-  return `${prefix}${path}`;
-}
-
-export function createProxy(server, method, path, config) {
-  const proxies = new Map([
-    ['/elasticsearch', server.plugins.elasticsearch.getCluster('data')],
-  ]);
-
-  const responseHandler = function (err, upstreamResponse) {
-    if (err) {
-      throw err;
-    }
-
-    if (upstreamResponse.headers.location) {
-      // TODO: Workaround for #8705 until hapi has been updated to >= 15.0.0
-      upstreamResponse.headers.location = encodeURI(upstreamResponse.headers.location);
-    }
-
-    return upstreamResponse;
-  };
-
-  for (const [proxyPrefix, cluster] of proxies) {
-    const options = {
-      method,
-      path: createPath(proxyPrefix, path),
-      config: {
-        timeout: {
-          socket: cluster.getRequestTimeout()
-        }
-      },
-      handler: {
-        proxy: {
-          mapUri: mapUri(cluster, proxyPrefix),
-          agent: createAgent({
-            url: cluster.getUrl(),
-            ssl: cluster.getSsl()
-          }),
-          xforward: true,
-          timeout: cluster.getRequestTimeout(),
-          onResponse: responseHandler
+  server.route({
+    method: 'POST',
+    path: '/elasticsearch/_msearch',
+    config: {
+      payload: {
+        parse: 'gunzip'
+      }
+    },
+    async handler(req, h) {
+      try {
+        const { payload } = req;
+        const body = payload
+          .toString('utf8')
+          .split('\n')
+          .filter(Boolean)
+          .map(JSON.parse);
+        const response = await callWithRequest(req, 'msearch', {
+          body
+        });
+        return h.response(response);
+      } catch(e) {
+        if (e instanceof Error) {
```

Comment:
This seems like a pretty wide net to catch and assume is a parse error. Might be better to have a try/catch block just for the parsing bit and a separate one for the callWithRequest call.

Reply:
👍 3d1f707